json/json.go

```diff
@@ -115,9 +116,21 @@ func (c *decoder) Decode(v interface{}) error {
 		if err != nil {
 			return err
 		}
+		return json.NewDecoder(r).Decode(v)
 	}
 
-	return json.NewDecoder(r).Decode(v)
+	decoder := json.NewDecoder(r)
+	err := decoder.Decode(v)
+	// the JSON decoder may read ahead so reset the reader so we don't
+	// drop the extra bytes read
+	c.r = decoder.Buffered()
```

Comment:
That won't work. That reader will only read out the rest of the buffered data. You'd need a MultiReader but then you'd end up with a recursive reader definition.

I think a better solution is to just not throw away the decoder. Using `io.MultiReader(c.jdecoder.Buffered(), c.r)` is probably sufficient. However, given some manual work, there may be more efficient ways to do this.

Reply:
> That reader will only read out the rest of the buffered data.

Oops, the tests seam to pass though.  Will fix.